File: dama-java/Dama/Dama/src/dama/Player.java
package dama;

import java.util.ArrayList;

public class Player {

    public static final int[][] NEIGHBOUR_OFFSETS = {{-1, -1}, {-1, 1}, {1, 1}, {1, -1}};

    protected Board board;  // the game board that the player is playing on
    private final boolean isTop;  // player's position
    private final String name; // the name of the player
    private static boolean isEgregna = true; // Egregna 

    // constructor
    public Player(Board board, boolean isTop, String name){
        this.board = board;
        this.isTop = isTop;
        this.name = name;
    }
    
    // constructor
    public Player(Board board, boolean isTop){
        this(board, isTop, isTop ? "Top Player" : "Bottom Player");
    }
    
    public String getName(){
        return name;
    }
    
    public static void setIsEgregna(boolean isItEgregna){
        isEgregna = isItEgregna;
    }
    
    
    // calculates the number of corkis that a player has left
    // and returns  the number of corkis that a player has left on the board
    public int getNumCorkis() {
        int numCorkis = 0;
        for (int row = 0; row < Board.NUM_ROWS; row++) {
            for (int col = 0; col < Board.NUM_ROWS; col++ ) {
                if (isTop) {
                    numCorkis += board.isTop(row, col) ? 1 : 0;
                } else {
                    numCorkis += board.isBottom(row, col) ? 1 : 0;
                }
            }
        }
        return numCorkis;
    }
    
    // identifies the legal cells(eats and move destinations) that can be reached out from a given
    // corki and returns them wrappen in an array list of arrays, the array at index 0 containing
    // all the moves and the array at index 1 containing all the eats
    public ArrayList<int[]>[] movesAndEats(int row, int col)  {
        int[] curCell = {row, col};
        ArrayList<int[]>[] movesAndEats = new ArrayList[]{new ArrayList<>(), new ArrayList<>()};
        for(int[] offset: NEIGHBOUR_OFFSETS){
            int[] neighborCell = {row + offset[0], col + offset[1]};
            if (board.isInsideBoard(neighborCell)) {
                if (canMove(curCell, neighborCell)) {
                    movesAndEats[0].add(neighborCell);
                } else {
                    if (canEat(curCell, neighborCell)) {
                        int[] landingCell = board.getNextCell(curCell, neighborCell);
                        movesAndEats[1].add(landingCell);
                    }
                }
            }
        }
        return movesAndEats;
    }

    public ArrayList<int[]>[] movesAndEats(int[] cell)  {
        return movesAndEats(cell[0], cell[1]);
    }

    public ArrayList<int[]> getCanAdvanceCells() {
        ArrayList<int[]> canAdvanceCells = new ArrayList();
        for (int row = 0; row < Board.NUM_ROWS; row++) {
            for (int col = 0; col < Board.NUM_ROWS; col++) {
                if ((isTop && board.isTop(row, col) || ((!isTop) && board.isBottom(row, col)))) {
                    ArrayList[] movesAndEats = movesAndEats(row, col);
                    int numMoves = movesAndEats[0].size();
                    int numEats = movesAndEats[1].size();
                    if (numMoves > 0 || numEats > 0) {
                        int[] curCell = {row, col};
                        canAdvanceCells.add(curCell);
                    }
                }
            }
        }
        return canAdvanceCells;
    }

    void playOneTurn(int[] startCell, int[] landingCell, boolean isAnEatMove)  {
        if(isAnEatMove){           
            int[] target = (!Player.isEgregna) && board.isKing(startCell) ? 
                    board.getTankegnaKingsPrey(startCell, landingCell) : 
                    board.getMiddleCell(startCell, landingCell);
            board.makeEmpty(target);
        }
        board.moveCorki(startCell, landingCell);        
        if (board.shouldKingify(landingCell)) { // crown the new king, if there is one
            board.makeKing(landingCell);
        }
    }
    
    // decides 
    void playTurn(int[][] advance){
        int[] fromCell = advance[0]; int[] landingCell = advance[1];
        if(canEat(fromCell, landingCell)){
            playOneTurn(fromCell, landingCell, true);            
        } else {
            playOneTurn(fromCell, landingCell, false);
        }
    }


    public boolean canMove(int[] fromCell, int[] toCell)  {
        if(!isOwnCorki(fromCell)) {
            System.out.println("NOT YOUR CORKI");
            return false;
        }
        if((!Player.isEgregna) && board.isKing(fromCell)) {
            return board.isEmptyInBetween(fromCell, toCell);
        } else {
            if(!board.areDiagonals(fromCell, toCell, 1)) {
                return false;
            }
            return board.isEmptyLegalCell(toCell) &&
                    (board.isKing(fromCell) || 
                    board.isForwardToTeter(fromCell, toCell));            
        }
    }

    
    public boolean canEat(int[] fromCell, int[] landingCell){
        if(!isOwnCorki(fromCell))
            return false;        
        if((!Player.isEgregna) && board.isKing(fromCell)) {
            if(!board.areDiagonals(fromCell, landingCell))
                return false;
            else {
                int[] prey = board.getTankegnaKingsPrey(fromCell, landingCell);
                return board.isEmptyLegalCell(landingCell) && prey != null;
            }            
        } else {
            if(!board.areDiagonals(fromCell, landingCell, 2))
                return false;
            else {
                int[] nextCell = board.getMiddleCell(fromCell, landingCell);
                return canEatHelper(fromCell, nextCell, landingCell);
            }            
        }        
                   
    }
    
    public boolean canEatHelper(int[] fromCell, int[] nextCell, int[] landingCell)  {
        if (board.areTeamMates(fromCell, nextCell) || 
                !board.areDiagonals(fromCell, nextCell, 1)) {
            return false;
        } else {
            boolean landingCellOk = landingCell != null &&
                    board.isEmptyLegalCell(landingCell);
            boolean isEatAllowed;
            if (Player.isEgregna) {
                isEatAllowed = (board.isKing(fromCell) && board.isOccupied(nextCell)) ||
                    (board.isTeter(nextCell) && board.isForwardToTeter(fromCell, nextCell));
            } else { // tankegna Teter
                isEatAllowed = board.isOccupied(nextCell);
            }
            return landingCellOk && isEatAllowed;
        }
    }    
    
    public boolean isOwnCorki(int row, int col){
        return (isTop && board.isTop(row, col)) ||
               ((! isTop) && board.isBottom(row, col));
    }
    
    public boolean isOwnCorki(int[] cell){
        return isOwnCorki(cell[0], cell[1]);
    }    
    
}

File: dama-py-tinkering/mess/player.py
"""
==================================================
THE Player IS A PROGRAM MODEL OF A DAMA PLAYER
==================================================
"""

NEIGHBOR_OFFSETS = ((-1, -1), (-1, 1), (1, 1), (1, -1))

class Player:
        
    def __init__(self, dama_board, is_top):
        """ creates a player object """
        self.board = dama_board
        self.is_top = is_top
        
    def getNumCorkis(self):
        """
        returns the number of corkis that a player have left on board
        """
        num_corkis = 0
        for row in range(self.board.NUM_ROWS):
            for col in range(self.board.NUM_ROWS):
                if self.is_top:
                    num_corkis += (self.is_top and self.board.isTop(row, col))
                else:
                    num_corkis += ((not self.is_top) and self.board.isBottom(row, col))
        return num_corkis
    
    def movesAndEats(self, row, col = None):
        """
        returns the cells a given corki@cur_cell can legally
        move to in one step, in Egregna Dama game.
        """
        if col is None: # row itself is [row, col]
            cur_cell = row            
        else:
            cur_cell = (row, col)
        legal_move_cells, legal_eat_cells = [], []
        for offset in NEIGHBOR_OFFSETS:
            neighbor_cell = cur_cell[0] + offset[0], cur_cell[1] + offset[1]
            if self.board.isInsideBoard(neighbor_cell):           
                if self.canMove(cur_cell, neighbor_cell):
                    legal_move_cells.append(neighbor_cell)
                else:
                    if self.canEat(cur_cell, neighbor_cell):
                        landing_cell = self.board.getNextCell(cur_cell, neighbor_cell)                      
                        legal_eat_cells.append(landing_cell)
        return legal_move_cells, legal_eat_cells

    def canMove(self, from_cell, to_cell):
        """
        returns true if corki at from_cell can move to the to_cell
        """
        return self.board.isEmptyLegalCell(to_cell) and \
               (self.board.isKing(from_cell) or
                self.board.isForwardToTeter(from_cell, to_cell))

    def canEat(self, from_cell, next_cell):
        """
        returns true if corki at from_cell can eat corki at to_cell
        """
        if self.board.areTeamMates(from_cell, next_cell):
            return False
        else:
            landing_cell = self.board.getNextCell(from_cell, next_cell)
            landing_cell_ok = self.board.isInsideBoard(landing_cell) and \
                              self.board.isEmptyLegalCell(landing_cell)
            is_eat_allowed = self.board.isKing(from_cell) or \
                              ( self.board.isTeter(from_cell) and
                                self.board.isForwardToTeter(from_cell, next_cell) )
            return landing_cell_ok and is_eat_allowed

    def movesAndEatsCombined(self, cell):
        """
        returns the next legal moves from cell
        """
        moves_and_eats = self.movesAndEats(cell)
        moves_and_eats[0].extend(moves_and_eats[1])
        return set(moves_and_eats[0])

    def getCanAdvanceCells(self):
        """
        returns cells whose corki can either eat or move
        """
        can_advance_cells = []
        for row in range(self.board.NUM_ROWS):
            for col in range(self.board.NUM_ROWS):
                if (self.is_top and self.board.isTop(row, col)) or \
                        ((not self.is_top) and self.board.isBottom(row, col)):
                    moves_and_eats = self.movesAndEats(row, col)
                    num_moves = len(moves_and_eats[0])
                    num_eats = len(moves_and_eats[1])
                    if (num_moves > 0) or (num_eats > 0):
                        can_advance_cells.append((row, col))
        return can_advance_cells
    
    def playOneTurn(self, start_cell, landing_cell, isAnEatMove):
        """ 
        performs dama move from start_cell to landing_cell,
        returns is_kingified
        """
        ''' perform move '''
        self.board.moveCorki(start_cell, landing_cell)
        ''' eat, if it is an eat move '''
        if isAnEatMove:
            mid_cell = self.board.getMiddleCell(start_cell, landing_cell)
            self.board.makeEmpty(mid_cell)
        ''' kingify if necessary '''
        is_kingified = False
        if self.board.shouldKingify(landing_cell):
            self.board.makeKing(landing_cell)
            is_kingified = True
        return is_kingified

class HumanPlayer(Player):
    
    def __init__(self, dama_board):
        super().__init__(dama_board, is_top = False)

    def playOneTurn(self, start_cell, landing_cell, isAnEatMove=False):
        """ 
        performs dama move from start_cell to landing_cell,
        returns is_kingified, isAnEatMove, (startcell, endcell)
        """ 
        moves_and_eats = self.movesAndEats(start_cell[0], start_cell[1])
        isAnEatMove = len(moves_and_eats[1]) > 0
        is_kingified = super().playOneTurn(start_cell, landing_cell, isAnEatMove)
        return is_kingified, isAnEatMove 
